sources/Adapters/picoTracker/filesystem/picoTrackerFileSystem.cpp

```diff
@@ -3,6 +3,123 @@
 #include "Application/Utils/wildcard.h"
 #include "System/Console/Trace.h"
 
+#include <string>
+
+// Max filename is actually 256 per FAT std
+static const int MAX_FILENAME_LEN = 128;
+
+picoTrackerPagedDir::picoTrackerPagedDir(const char *path)
+    : path_{std::string(path)} {
+  fileIndexes_.reserve(354);
+  subdirIndexes_.reserve(96);
+};
+
+void picoTrackerPagedDir::GetContent(const char *mask) {
+  Trace::Log("PAGEDFILESYSTEM", "GetContent path:%s mask:%s", path_.c_str(),
+             mask);
+  fileIndexes_.clear();
+  subdirIndexes_.clear();
+  FsBaseFile dir;
+
+  if (!dir.open(path_.c_str())) {
+    Trace::Error("PagedDir GetContent Failed to open %s", path_.c_str());
+    return;
+  }
+
+  if (!dir.isDir()) {
+    Trace::Error("Path:%s is not a directory", path_.c_str());
+    return;
+  }
+
+  int count = 0;
+  FsBaseFile entry;
+  while (entry.openNext(&dir, O_READ)) {
+    char current[MAX_FILENAME_SIZE];
+    entry.getName(current, MAX_FILENAME_SIZE);
+
+    int fileIndex = entry.dirIndex();
+    if (entry.isDir()) {
+      subdirIndexes_.push_back(fileIndex);
+    } else if (wildcardfit(mask, current)) {
+      fileIndexes_.push_back(fileIndex);
+    }
+    count++;
+  }
+  Trace::Log("PAGEDFILESYSTEM", "scanned %d files", count);
+}
+
+std::string picoTrackerPagedDir::getFullName(int index) {
+  FsBaseFile dir;
+  if (!dir.open(path_.c_str())) {
+    Trace::Error("getFileList failed open:%s", path_.c_str());
+    return std::string("");
+  }
+  if (!dir.isDir()) {
+    Trace::Error("Path:%s is not a directory", path_.c_str());
+    return std::string("");
+  }
+  char filename[MAX_FILENAME_LEN];
+  FsBaseFile file;
+
+  if (!file.open(&dir, index, O_READ)) {
+    Trace::Error("PAGEDFILESYSTEM Failed to getfile at Index %d", index);
+  }
+  file.getName(filename, MAX_FILENAME_LEN);
+  return std::string(filename);
+}
+
+void picoTrackerPagedDir::getFileList(int startOffset,
+                                      std::vector<FileListItem> *fileList) {
+  bool addedParentDirEntry = false;
+  FsBaseFile dir;
+
+  if (!dir.open(path_.c_str())) {
+    Trace::Error("getFileList failed open:%s", path_.c_str());
+    return;
+  }
+  if (!dir.isDir()) {
+    Trace::Error("Path:%s is not a directory", path_.c_str());
+    return;
+  }
+  static const int MAX_ITEMS = 15;
```

Comment:
I have a (bad?) habit of trying not to use the preprocessor, but yes it could be a define though I think it would basically come out as the same generated code I think, but I can change it to a define if you prefer that?